### PR TITLE
Feature/sbachmei/mic 4994 write final results

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*:  bugfix, refactor
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-4968

### Changes and notes
 
This PR updates the .report() method to actually save out the
final results instead of the wide output.hdf. It also properly handles
the edge case where there are no stratifications and adds a test
against that.

I'd highly recommend reviewing each of the two commits separately!

### Testing
 tests pass

### CODE REVIEW

File: src/vivarium/framework/engine.py [hunk 1]
```diff
@@ -274,8 +270,22 @@ def report(self, print_results: bool = True) -> Dict[str, Any]:
                 float_format=lambda x: f"{x:.2f}",
             )
             self._logger.info("\n" + performance_metrics)
+        for measure, df in metrics.items():
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:273]: What if each observation was also registered with a `report` method that was called on the `simulation_end` event? For all observers that we have so far it would just be what you have here, but it would enable additional flexibility if we had an observer that wanted to write results in a different way (e.g. to a parquet or hdf file).

Comment (stevebachmeier) [reply src/vivarium/framework/engine.py:273]: Hmm, that's a good idea - I'll implement and see how it goes

File: src/vivarium/framework/results/context.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def add_observation(
 
     def gather_results(
         self, population: pd.DataFrame, event_name: str
-    ) -> Generator[Optional[pd.DataFrame], None, None]:
+    ) -> Generator[Tuple[Optional[pd.DataFrame], Optional[str]], None, None]:
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/context.py:106]: Rajan: Try yielding only None w/ typing `-> Generator[Optional[Tuple[pd.Dataframe, str]], None, None]`

File: src/vivarium/framework/results/context.py [hunk 3]
```diff
@@ -132,9 +133,19 @@ def gather_results(
...
                     else:
-                        aggregates = pop_groups.apply(aggregator)
-
-                    # Ensure we are dealing with a single column of formattable results
+                        # aggregate all columns
+                        if (
+                            isinstance(pop_groups, pd.DataFrame)
+                            and len(pop_groups.columns) > 1
+                        ):
+                            # Apply the aggregator to the entire group
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/context.py:141]: Add comment that this only happens if there are no stratifications

File: src/vivarium/framework/results/context.py [hunk 4]
```diff
@@ -119,10 +119,11 @@ def gather_results(
             else:
                 filtered_pop = population
             if filtered_pop.empty:
-                yield None
+                yield None, None
             else:
-                if not list(stratifications):  # Handle situation of no stratifications
-                    pop_groups = filtered_pop.groupby(lambda _: True)
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/context.py:125]: Rajan: let's go back to making a pop_group of everything - even no stratificatons. Call it "all" or whatever.

File: src/vivarium/framework/results/context.py [hunk 5]
```diff
@@ -132,9 +133,19 @@ def gather_results(
                             pop_groups[aggregator_sources].apply(aggregator).fillna(0.0)
                         )
                     else:
-                        aggregates = pop_groups.apply(aggregator)
-
-                    # Ensure we are dealing with a single column of formattable results
+                        # aggregate all columns
+                        if (
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/context.py:137]: This all goes away when everything is a groupby object

File: src/vivarium/framework/results/manager.py [hunk 6]
```diff
@@ -87,16 +87,20 @@ def on_post_setup(self, _: Event):
...
-                    # Get the complete cartesian product of stratifications
-                    index_values = list(itertools.product(*stratification_values.values()))
+                    if stratification_values:
+                        # Create index of the complete stratifications
+                        idx = pd.MultiIndex.from_tuples(
+                            list(itertools.product(*stratification_values.values())),
+                            names=stratification_values.keys(),
+                        )
+                    else:
+                        idx = pd.Index(aggregator_sources, name=measure)
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/manager.py:97]: Rajan: hopefully this case goes away when we go back to using groupby objects (even for no stratifications)

File: src/vivarium/framework/results/manager.py [hunk 7]
```diff
@@ -257,9 +263,8 @@ def _prepare_population(self, event: Event):
             population[pipeline.name] = pipeline(event.index)
         return population
 
-    def get_results(self, index, metrics):
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/manager.py:260]: Add typing to some of this stuff

File: src/vivarium/interface/cli.py [hunk 8]
```diff
@@ -143,13 +143,14 @@ def run(
...
     finished_sim = main(model_specification, configuration=override_configuration)
-
-    # TODO [MIC-4994] finished_sim.report(results_root)
-    metrics = pd.DataFrame(finished_sim.report(), index=[0])
-    metrics["simulation_run_time"] = time() - start
-    metrics["random_seed"] = finished_sim.configuration.randomness.random_seed
-    metrics["input_draw"] = finished_sim.configuration.input_data.input_draw_number
-    metrics.to_hdf(results_root / "output.hdf", key="data")
+    finished_sim.report(results_root)
+    breakpoint()
```
Comment (stevebachmeier) [inline src/vivarium/interface/cli.py:147]: remove

File: tests/framework/results/test_context.py [hunk 9]
```diff
@@ -264,9 +264,10 @@ def test_gather_results(
     )
 
     i = 0
-    for r in ctx.gather_results(population, event_name):
```
Comment (stevebachmeier) [inline tests/framework/results/test_context.py:267]: change _measure to _

File: tests/framework/test_engine.py [hunk 10]
```diff
@@ -243,14 +258,114 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+        HousePointsObserver(),
+        NoStratificationsQuidditchWinsObserver(),
+        QuidditchWinsObserver(),
+        HogwartsResultsStratifier(),
+    ]
+    finished_sim = run_simulation(components=components, configuration=CONFIG)
+    finished_sim.report(results_root)
+
+    # The observers are based on number of steps - extract how many steps were run
+    time_dict = finished_sim.configuration.time.to_dict()
```
Comment (stevebachmeier) [inline tests/framework/test_engine.py:290]: Rajan: consider updating the plugin to use SimpleClock because it just uses timesteps (not datetime)

File: tests/framework/test_engine.py [hunk 11]
```diff
@@ -243,14 +258,114 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+    end_date = _convert_to_datetime(time_dict["end"])
+    start_date = _convert_to_datetime(time_dict["start"])
+    num_steps_float = (end_date - start_date).days / time_dict["step_size"]
+    # round up
+    num_steps = int(num_steps_float) + 1 if num_steps_float % 1 else num_steps_float
+
+    # Check for expected results and confirm format
+    results_list = [file for file in results_root.rglob("*")]
+    assert len(results_list) == 3
```
Comment (stevebachmeier) [inline tests/framework/test_engine.py:299]: don't need this since w'ere checking names below

File: tests/framework/test_engine.py [hunk 12]
```diff
@@ -243,14 +258,114 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+    # (or, when no stratifications, just a single "all" row)
+    assert set(zip(house_points["student_house"], house_points["power_level"])) == set(
+        product(STUDENT_HOUSES, POWER_LEVELS)
+    )
+    assert set(zip(quidditch_wins["familiar"], quidditch_wins["power_level"])) == set(
+        product(FAMILIARS, POWER_LEVELS)
+    )
+
+    # Check that all values are 0 except for expected groups
+    assert (
```
Comment (stevebachmeier) [inline tests/framework/test_engine.py:325]: Reformat this so it's more readable (black made it tough to read)

File: tests/framework/test_engine.py [hunk 13]
```diff
@@ -243,14 +258,114 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+            "value",
+        ]
+        == 0
+    ).all()
+    assert (
+        quidditch_wins.loc[(quidditch_wins["familiar"] != "banana_slug"), "value"] == 0
+    ).all()
+
+    # Check that expected groups' values are a multiple of the number of steps
+    assert (
```
Comment (stevebachmeier) [inline tests/framework/test_engine.py:340]: Add a check that they are NOT 0 (0 % 0 == 0)

File: tests/framework/test_engine.py [hunk 14]
```diff
@@ -243,14 +258,114 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+        ]
+        % num_steps
+        == 0
+    ).all()
+    assert (
+        quidditch_wins.loc[quidditch_wins["familiar"] == "banana_slug", "value"] % num_steps
+        == 0
+    ).all()
+
+    # Check for other cols
```
Comment (stevebachmeier) [inline tests/framework/test_engine.py:354]: Do as much in a loop as possible

File: src/vivarium/framework/engine.py (hunk 1 above)
Comment (stevebachmeier) [reply src/vivarium/framework/engine.py:273]: As discussed, this will happen in a follow-up PR

File: src/vivarium/framework/results/context.py (hunk 2 above)
Comment (stevebachmeier) [reply src/vivarium/framework/results/context.py:106]: Not a clean way to do this

File: src/vivarium/framework/results/manager.py (hunk 6 above)
Comment (stevebachmeier) [reply src/vivarium/framework/results/manager.py:97]: It doesn't work
```
(Pdb++) pd.MultiIndex.from_tuples([()], names=[])
*** ValueError: Length of names must match number of levels in MultiIndex.
Traceback (most recent call last):
  File "/ihme/homes/sbachmei/miniconda3/envs/dev-results-processing/lib/python3.11/site-packages/pandas/core/indexes/multi.py", line 222, in new_meth
    return meth(self_or_cls, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/ihme/homes/sbachmei/miniconda3/envs/dev-results-processing/lib/python3.11/site-packages/pandas/core/indexes/multi.py", line 598, in from_tuples
    return cls(
           ^^^^
  File "/ihme/homes/sbachmei/miniconda3/envs/dev-results-processing/lib/python3.11/site-packages/pandas/core/indexes/multi.py", line 356, in __new__
    result._set_names(names)
  File "/ihme/homes/sbachmei/miniconda3/envs/dev-results-processing/lib/python3.11/site-packages/pandas/core/indexes/multi.py", line 1609, in _set_names
    raise ValueError(
```

File: src/vivarium/framework/results/context.py (hunk 4 above)
Comment (stevebachmeier) [reply src/vivarium/framework/results/context.py:125]: ✅

File: src/vivarium/framework/results/manager.py (hunk 7 above)
Comment (stevebachmeier) [reply src/vivarium/framework/results/manager.py:260]: Will do in a separate PR

File: src/vivarium/interface/cli.py (hunk 8 above)
Comment (stevebachmeier) [reply src/vivarium/interface/cli.py:147]: ✅

File: tests/framework/test_engine.py (hunk 11 above)
Comment (stevebachmeier) [reply tests/framework/test_engine.py:299]: ✅

File: tests/framework/test_engine.py (hunk 12 above)
Comment (stevebachmeier) [reply tests/framework/test_engine.py:325]: ✅

File: tests/framework/test_engine.py (hunk 13 above)
Comment (stevebachmeier) [reply tests/framework/test_engine.py:340]: ✅

File: tests/framework/test_engine.py (hunk 14 above)
Comment (stevebachmeier) [reply tests/framework/test_engine.py]: ✅

File: src/vivarium/framework/results/context.py [hunk 15]
```diff
@@ -134,14 +138,15 @@ def gather_results(
                     else:
                         aggregates = pop_groups.apply(aggregator)
 
-                    # Ensure we are dealing with a single column of formattable results
+                    # Ensure we are dealing with a single column of formattable results):
```
Comment (rmudambi) [inline src/vivarium/framework/results/context.py:141]: Typo?

File: tests/framework/results/test_manager.py [hunk 16]
```diff
@@ -301,6 +301,21 @@ def test_stratified_metrics_initialized_as_zeros_dataframes():
     )
 
 
+def test_no_stratified_metrics_initialization():
+    """Test that if no stratifications are registered then we initialize a
+    single-row DataFrame with 'value' of zero and index labeled 'all'
+    """
+    components = [QuidditchWinsObserver(), HousePointsObserver(), Hogwarts()]
```
Comment (rmudambi) [inline tests/framework/results/test_manager.py:308]: Shouldn't we get an error, since we have observers that are asking for stratifications that haven't been registered?

**Review by rmudambi — APPROVED** [review]
Approved, though I do have one question about a test.

File: tests/framework/results/test_manager.py (hunk 16 above)
Comment (stevebachmeier) [reply tests/framework/results/test_manager.py:308]: Hmm, I don't think that's a current feature. There's no guarantee at the point of observer registration what stratifications will eventually exist (I think?). Is this something we should be supporting? It could happen in post setup or something.

File: tests/framework/test_engine.py [hunk 17]
```diff
@@ -243,14 +257,105 @@ def test_SimulationContext_finalize(SimulationContext, base_config, components):
...
+    finished_sim = run_simulation(components=components, configuration=CONFIG)
+    finished_sim.report(results_root)
+
+    # The observers are based on number of steps - extract how many steps were run
+    time_dict = finished_sim.configuration.time.to_dict()
+    end_date = _convert_to_datetime(time_dict["end"])
+    start_date = _convert_to_datetime(time_dict["start"])
+    num_steps_float = (end_date - start_date).days / time_dict["step_size"]
+    # round up
+    num_steps = int(num_steps_float) + 1 if num_steps_float % 1 else num_steps_float
```
Comment (patricktnast) [inline tests/framework/test_engine.py:295]: nit: the `else` makes this return a float  (e.g. 5.0) instead of 5. Is that what you want? Does math.ceil() not work?

File: tests/framework/results/test_manager.py (hunk 16 above)
Comment (rmudambi) [reply tests/framework/results/test_manager.py:308]: Fair enough.

**Review by patricktnast — APPROVED** [review]
LGTM

File: tests/framework/test_engine.py (hunk 17 above)
Comment (stevebachmeier) [reply tests/framework/test_engine.py:295]: Eh, I was honestly just too lazy to go to the top to import a new library. I'll use `math.ceil` though; it's more readable
